front/components/agent_builder/observability/ToolLatencyChart.tsx

```diff
@@ -0,0 +1,209 @@
+  const versionData = toolLatencyByVersion.find((v) => v.version === version);
+
+  const rows = payload
+    .filter((p) => typeof p.value === "number" && p.value > 0)
+    .sort((a, b) => b.value - a.value)
+    .flatMap((p) => {
+      const toolName = p.name || "";
+      const toolData = versionData?.tools[toolName];
+
+      if (!toolData) {
+        return [
+          {
+            label: toolName,
+            value: `${p.value}ms`,
+            colorClassName: getToolColor(toolName, topTools),
+          },
+        ];
+      }
+
+      return [
+        {
+          label: `${toolName} (avg)`,
+          value: `${toolData.avgLatencyMs}ms`,
+          colorClassName: getToolColor(toolName, topTools),
+        },
+        {
+          label: `${toolName} (p50)`,
+          value: `${toolData.p50LatencyMs}ms`,
+          colorClassName: getToolColor(toolName, topTools),
+        },
+        {
+          label: `${toolName} (p95)`,
+          value: `${toolData.p95LatencyMs}ms`,
+          colorClassName: getToolColor(toolName, topTools),
+        },
+      ];
+    });
+
+  return <ChartTooltipCard title={String(label)} rows={rows} />;
+}
+
+export function ToolLatencyChart({
+  workspaceId,
+  agentConfigurationId,
+}: ToolLatencyChartProps) {
+  const [period, setPeriod] =
+    useState<ObservabilityTimeRangeType>(DEFAULT_PERIOD_DAYS);
+
+  const { toolLatencyByVersion, isToolLatencyLoading, isToolLatencyError } =
+    useAgentToolLatency({
+      workspaceId,
+      agentConfigurationId,
+      days: period,
+      disabled: !workspaceId || !agentConfigurationId,
+    });
+
+  const { chartData, topTools } = useMemo(() => {
+    if (!toolLatencyByVersion || toolLatencyByVersion.length === 0) {
+      return { chartData: [], topTools: [] };
+    }
+
+    // Calculate top tools by weighted average latency
+    const top = calculateTopTools(
+      toolLatencyByVersion,
+      (toolData) => toolData.avgLatencyMs * toolData.count,
+      MAX_TOOLS_DISPLAYED
+    );
+
+    // Build rows with average latency values for each tool
+    const rows: ChartRow[] = toolLatencyByVersion.map((v) => {
+      const toolValues: Record<string, number> = {};
+      for (const toolName of top) {
+        const t = v.tools[toolName];
+        toolValues[toolName] = t ? t.avgLatencyMs : 0;
+      }
+
+      return { version: `v${v.version}`, values: toolValues };
+    });
+
+    return { chartData: rows, topTools: top };
+  }, [toolLatencyByVersion]);
+
+  const renderTooltip = useCallback(
+    (props: TooltipContentProps<number, string>) => (
+      <ToolLatencyTooltip
+        {...props}
+        topTools={topTools}
+        toolLatencyByVersion={toolLatencyByVersion ?? []}
+      />
+    ),
+    [topTools, toolLatencyByVersion]
+  );
+
+  const legendItems = topTools.map((toolName) => ({
+    key: toolName,
+    label: toolName,
+    colorClassName: getToolColor(toolName, topTools),
+  }));
+
+  return (
+    <ChartContainer
```

Comment:
I feel like I saw the same code in your previous PR, anyway we can reuse the same component? 🤔

Reply:
Well we reuse the `ChartContainer` component right?